src/core/iter.rs

```diff
@@ -0,0 +1,108 @@
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Iterators over DAGs
+
+/// Structure that can be iterated like a DAG _(directed acyclic graph)_.
+pub trait DagIterable: Sized {
+    /// Node in the DAG
+    type Node: Copy;
+    /// Auxiliary data for internal computations
+    type Aux;
+
+    /// Return the DAG root, if the DAG is nonempty.
+    fn root(&self) -> Option<Self::Node>;
+
+    /// Return the left child of the given `node`.
+    fn left_of(&self, node: Self::Node) -> Option<Self::Node>;
+
+    /// Return the right child of the given `node`.
+    fn right_of(&self, node: Self::Node) -> Option<Self::Node>;
+
+    /// During post-order traversal in [PostOrderIter::new()],
+    /// return whether `left` or `right` are a cross-reference to nodes
+    /// that appear earlier in the order.
+    ///
+    /// If `maybe_right = Some(right)`, then `right` is checked.
+    /// Otherwise, `left` is checked.
+    ///
+    /// Arguments `to_visit` and `aux` provide additional data.
+    fn is_cross_reference(
+        &self,
+        node: Self::Node,
+        left: Self::Node,
+        maybe_right: Option<Self::Node>,
+        to_visit: &[Self::Node],
+        aux: &Self::Aux,
+    ) -> bool;
+
+    /// Return auxiliary data for internal computations.
+    fn get_aux(&self) -> Self::Aux;
+
+    /// Return a post-order iterator over the DAG.
+    fn iter_post_order(&self) -> PostOrderIter<Self> {
+        PostOrderIter::new(self)
+    }
+}
+
+/// Iterates over a DAG in _post order_.
+/// That means, left children appear before right ones, and children appear before their parent.
+/// Shared nodes appear only once at their leftmost position.
+pub struct PostOrderIter<D: DagIterable> {
+    stack: Vec<D::Node>,
+}
+
+impl<D: DagIterable> PostOrderIter<D> {
+    /// Create a new iterator from the given `dag`.
+    pub fn new(dag: &D) -> Self {
+        let mut to_visit = match dag.root() {
+            Some(node) => vec![node],
+            None => return PostOrderIter { stack: Vec::new() },
+        };
+        let mut visited = Vec::new();
+        let extra = dag.get_aux();
+
+        while let Some(node) = to_visit.pop() {
+            visited.push(node);
+
+            if let Some(left) = dag.left_of(node) {
+                if !dag.is_cross_reference(node, left, None, &to_visit, &extra) {
+                    to_visit.push(left);
+                }
+
+                if let Some(right) = dag.right_of(node) {
+                    if !dag.is_cross_reference(node, left, Some(right), &to_visit, &extra) {
+                        to_visit.push(right);
+                    }
+                }
+            }
+        }
+
+        PostOrderIter { stack: visited }
```

Comment:
In cf8832de5c3d800f6f9a63d9620f537450368908:

This whole thing should be in `PostOrderIter::next`. Then you could maintain `to_visit` in the iterator and not need the `visited` variable at all, saving space as well as a ton of time during iterator construction, and avoid wasting time/space for users who don't want to iterate through the whole thing.

Reply:
I think you misunderstand the algorithm: What `visited` contains is the nodes in the _reverse_ order _(AFAIU, this is a "reverse pre-order", adding to the confusion)_. We visit the parent, its right children and then its left children. This means that in order to get the first node in post-order, we need to visit all nodes. The fastest approach is to store all visited nodes in a stack (`visited`) and to pop from that.

At first glance, it seems wasteful to visit the entire graph in the iterator constructor. However, even for large programs (1k~10k nodes) this is fast thanks to the cross-reference checking. We could make it so that the graph is visited on the first call of `next()`, but I am not sure if this is worth it. In any case, not iterating over the entire graph is _impossible_.

Reply:
To check canonical order, we actually do not need post-order iterators. Reverse-pre-order iterators would work, too _(post order iff successor index is predecessor minus 1)_. In this case, we could write a lazy iterator that only visits as much of the graph as it needs to. Two thoughts regarding this:
(1) To convert `TermDag` to `UntypedProgram`, we need post order. Maybe, we don't want to write two iterators for very similar use cases.
(2) When checking for canonical order, we greedily visit all nodes anyway. A lazy iterator wouldn't add value for this use case.

Reply:
Right now you're using O(program size) space in this iterator. You should be able to use O(program depth) space.

I see now that this is a little more involved because you'll need to change your algorithm, but I don't see why it should be impossble to do a post-order traversal without first iterating (and storing) the entire program.

Reply:
I suggest that we postpone this optimisation to a future PR. What I used was a standard algorithm that might have sub-optimal space complexity. There might be other algorithms, but I didn't find them over the course of a week. More work seems like diminishing returns to me. We could open an issue.

Reply:
Actually, I take that back. Looking at the [algorithm on Wikipedia](https://en.wikipedia.org/w/index.php?title=Tree_traversal&oldid=1083604462#Post-order_implementation), it seems to do exactly what you wanted. It is O(depth) and lazy, and worth having. I implemented it before, and it seemed slow for some reason, but I will try it again.